static/js/tools/map/compute/map_value_dates.ts

```diff
@@ -0,0 +1,162 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This module contains custom React hooks that makes computation for map chart.
+
+import _ from "lodash";
+import { Dispatch, useContext, useEffect } from "react";
+
+import {
+  EntityObservation,
+  EntityObservationList,
+} from "../../../shared/stat_types";
+import { ChartDataType, ChartStore, ChartStoreAction } from "../chart_store";
+import { useIfRatio } from "../condition_hooks";
+import { Context } from "../context";
+import {
+  useAllStatReady,
+  useDefaultStatReady,
+  useDenomStatReady,
+  useGeoJsonReady,
+} from "../ready_hooks";
+import { DataPointMetadata, getPlaceChartData } from "../util";
+
+export function useComputeMapValueAndDate(
+  chartStore: ChartStore,
+  dispatchChartStore: Dispatch<ChartStoreAction>,
+  dispatchSources: Dispatch<Set<string>>,
+  dispatchMetadata: Dispatch<Record<string, DataPointMetadata>>
+) {
+  const { statVar, placeInfo, dateCtx } = useContext(Context);
+  const geoJsonReady = useGeoJsonReady(chartStore);
+  const allStatReady = useAllStatReady(chartStore);
+  const defaultStatReady = useDefaultStatReady(chartStore);
+  const denomStatReady = useDenomStatReady(chartStore);
+  const ifRatio = useIfRatio();
+  return useEffect(() => {
+    if (!geoJsonReady()) {
+      return;
+    }
+    if (statVar.value.metahash) {
+      if (!allStatReady()) {
+        return;
+      }
+    } else {
+      if (!defaultStatReady()) {
+        return;
+      }
+    }
+    if (
+      chartStore.mapValuesDates.context &&
+      dateCtx.value === chartStore.mapValuesDates.context.date &&
+      _.isEqual(statVar.value, chartStore.mapValuesDates.context.statVar) &&
+      _.isEqual(placeInfo.value, chartStore.mapValuesDates.context.placeInfo)
+    ) {
+      return;
+    }
+    if (ifRatio && !denomStatReady()) {
+      return;
+    }
+    console.log("[Map Compute] map values and dates");
+    const mapValues = {};
+    const sources = new Set<string>();
+    const mapDates = new Set<string>();
+    const metadata = {};
+    const facets = Object.assign(
+      [],
+      chartStore.defaultStat.data.facets,
+      chartStore.denomStat.data ? chartStore.denomStat.data.facets : {},
+      chartStore.allStat.data ? chartStore.allStat.data.facets : {}
+    );
+    for (const geoFeature of chartStore.geoJson.data.features) {
```

Comment:
I wonder if instead of looping through chartStore.geoJson.data.features, it would be better to just loop through the stat data. That way map values won't have to depend on geoJson